mkosi/run.py

```diff
@@ -0,0 +1,273 @@
+import contextlib
+import ctypes
+import ctypes.util
+import multiprocessing
+import os
+import pwd
+import shlex
+import signal
+import subprocess
+import sys
+from pathlib import Path
+from types import FrameType
+from typing import Any, Iterable, Iterator, Mapping, Optional, Sequence
+
+from mkosi.backend import MkosiState
+from mkosi.log import ARG_DEBUG, MkosiPrinter, die
+from mkosi.types import _FILE, CompletedProcess, PathString, Popen
+
+CLONE_NEWNS = 0x00020000
+CLONE_NEWUSER = 0x10000000
+
+
+def unshare(flags: int) -> None:
+    libc_name = ctypes.util.find_library("c")
+    if libc_name is None:
+        die("Could not find libc")
+    libc = ctypes.CDLL(libc_name, use_errno=True)
+
+    if libc.unshare(ctypes.c_int(flags)) != 0:
+        e = ctypes.get_errno()
+        raise OSError(e, os.strerror(e))
+
+
+def read_subrange(path: Path) -> tuple[int, int]:
+    for line in path.read_text().splitlines():
+        name, start, count = line.split(":")
+
+        if name == pwd.getpwuid(os.getuid()).pw_name:
+            break
+    else:
+        die(f"No mapping found for current user in {path}")
+
+    return int(start), int(count)
+
+
+def become_root(uid: int, gid: int) -> tuple[int, int]:
+    if os.getuid() == 0:
+        return uid, gid
+
+    subuid, subuidcount = read_subrange(Path("/etc/subuid"))
+    subgid, subgidcount = read_subrange(Path("/etc/subgid"))
+
+    event = multiprocessing.Event()
+    pid = os.getpid()
+
+    child = os.fork()
+    if child == 0:
+        event.wait()
+
+        # We map the private UID range configured in /etc/subuid and /etc/subgid into the container using
+        # newuidmap and newgidmap. On top of that, we also make sure to map in the user running mkosi so that
+        # we can run still chown stuff to that user or run stuff as that user which will make sure any
+        # generated files are owned by that user.
+
+        newuidmap = [
+            "newuidmap", pid,
+            0, subuid, subuidcount - 100,
```

Comment:
What if `subuidcount` is smaller than 100? If we disallow that, this should be caught earlier.

Reply:
I added enforcing that the range is of size 65536